src/appshell/view/preferences/generalpreferencesmodel.cpp

```diff
@@ -54,30 +55,27 @@ void GeneralPreferencesModel::load()
 
 void GeneralPreferencesModel::checkUpdateForCurrentLanguage()
 {
-    QString currentLanguageCode = this->currentLanguageCode();
-    LanguageStatus::Status languageStatus = languagesService()->languageStatus(currentLanguageCode);
-    if (languageStatus != LanguageStatus::Status::NeedUpdate) {
-        Language currentLanguage = languagesService()->languages().val.value(currentLanguageCode);
-        QString msg = mu::qtrc("appshell/preferences", "Your version of %1 is up to date").arg(currentLanguage.name);
-        interactive()->info(msg.toStdString(), "");
-        return;
-    }
+    QString languageCode = languagesService()->effectiveLanguageCode(currentLanguageCode());
 
-    RetCh<LanguageProgress> progress = languagesService()->update(currentLanguageCode);
-    if (!progress.ret) {
-        LOGE() << progress.ret.toString();
-        return;
-    }
+    m_languageUpdateProgress = languagesService()->update(languageCode);
+
+    m_languageUpdateProgress.progressChanged.onReceive(this, [this](int64_t current, int64_t total, const std::string& status) {
+        emit receivingUpdateForCurrentLanguage(current, total, QString::fromStdString(status));
+    });
 
-    progress.ch.onReceive(this, [this](const LanguageProgress& progress) {
-        emit receivingUpdateForCurrentLanguage(progress.current, progress.status);
-    }, Asyncable::AsyncMode::AsyncSetRepeat);
+    m_languageUpdateProgress.finished.onReceive(this, [this, languageCode](Ret ret) {
+        if (ret.code() == static_cast<int>(Err::AlreadyUpToDate)) {
+            QString msg = mu::qtrc("appshell/preferences", "Your version of %1 is up to date.")
+                          .arg(languagesService()->languages()[languageCode].name);
+            interactive()->info(msg.toStdString(), "");
+        }
+    });
 }
 
 QVariantList GeneralPreferencesModel::languages() const
 {
-    ValCh<LanguagesHash> languages = languagesService()->languages();
-    QList<Language> languageList = languages.val.values();
+    LanguagesHash languages = languagesService()->languages();
```

Comment:
`const LanguagesHash&. 
`

Also, ideally, this line should be removed. We can iterate over the container using iterators
`QList<Language> languageList = languages.values(); `

Reply:
There are two possibilities:
```c++
    QList<Language> languages = languagesService()->languages().values();

    std::sort(languages.begin(), languages.end(), [](const Language& l, const Language& r) {
        return l.code < r.code;
    });

    QVariantList result;

    for (const Language& language : languages) {
        QVariantMap languageObj;
        languageObj["code"] = language.code;
        languageObj["name"] = language.name;
        result << languageObj;
    }
```

```c++
    QVariantList result;

    for (const Language& language : languagesService()->languages()) {
        QVariantMap languageObj;
        languageObj["code"] = language.code;
        languageObj["name"] = language.name;
        result << languageObj;
    }

    std::sort(result.begin(), result.end(), [](const QVariant& l, const QVariant& r) {
        return l.toMap().value("code").toString() < r.toMap().value("code").toString();
    });
```

In the first version, the sorting is probably more efficient, but the `.values()` is less efficient. 
In the second version, the iteration is faster but the sorting is more cumbersome. (Apparently, if you iterate over a QHash in `for (… : …)` style, you get the values only, not a key+value pair like with std::map.)

I've a light preference for the first version, what's your opinion?